src/blob.jl

```diff
@@ -63,36 +63,54 @@ The number of bytes needed to allocate `T` itself.
 
 Defaults to `sizeof(T)`.
 """
-@generated function self_size(::Type{T}) where T
-    @assert isconcretetype(T)
+Base.@assume_effects :foldable function self_size(::Type{T}) where T
+    # This function is marked :foldable to encourage constant folding for this types-only
+    # static computation.
     if isempty(fieldnames(T))
-        quote
-            $(Expr(:meta, :inline))
```

Comment:
Do we retain any guarantees of inlining the result of self_size without the inline meta? 🤔

Reply:
discussed on zoom. We think there is significant overlap in effect between inlining and const-propagating, and we decided to prefer the const-prop because of the recursive nature of this function.